src/proof.rs

```diff
@@ -597,16 +595,41 @@ impl<'a, 'b> ProofTreePrinterImpl<'a, 'b> {
         }
         process_block(&mut paren_stmt, &mut length_block);
 
+        self.write_word("(")?;
+        for s in &paren_stmt {
+            self.write_word(as_str(s.label()))?;
+        }
+        self.write_word(")")?;
+
+        let mut ess_stmt: Vec<StatementRef> = frame
+            .hypotheses
+            .iter()
+            .filter_map(|h| {
+                if let Hyp::Essential(addr, _) = h {
+                    Some(self.p.sset.statement(*addr))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        ess_stmt.append(&mut paren_stmt);
+        paren_stmt = ess_stmt;
+
         let mut letters: Vec<u8> = vec![];
         for item in &rpn {
             let (is_fwdref, mut letter) = match *item {
                 RPNStep::Normal { fwdref, addr, .. } => {
                     let stmt = self.p.sset.statement(addr);
                     let pos = if stmt.statement_type() == Floating {
-                        let atom = self.p.nset.var_atom(stmt).unwrap();
-                        frame.var_list[..frame.mandatory_count]
```

Comment:
`var_list` does not appear to list the mandatory floating hypotheses in the right order. I've replaced it with a lookup in the frame as built in the scope pass.